File: .origin/microfilm/export_service.py
"""
Export service module for handling microfilm export operations.

This module provides a service layer for export-related operations,
including exporting project data, document information, film allocations,
film rolls, and index data to JSON files.
"""

import os
import json
from pathlib import Path
from typing import Dict, Any, Optional

from models import Project, FilmAllocation

class ExportService:
    """
    Service for handling export operations.
    
    This service handles exporting project data, document information,
    film allocations, and other relevant data to JSON files.
    """
    
    def __init__(self, logger=None):
        """
        Initialize the export service.
        
        Args:
            logger: Optional logger instance.
        """
        self.logger = logger
    
    def export_results(self, project: Project, index_data: Optional[Dict[str, Any]] = None) -> Dict[str, Path]:
        """
        Export project results to JSON files.
        
        Args:
            project: Project with processed documents and film allocation
            index_data: Optional index data dictionary
            
        Returns:
            Dictionary with paths to exported files
        """
        if self.logger:
            self.logger.section("Exporting Results")
            self.logger.film_info("Starting export of project results")
        
        result_paths = {}
        
        try:
            # Create a data directory in the project path
            data_dir = project.project_path / ".data"
            data_dir.mkdir(exist_ok=True)
            
            if self.logger:
                self.logger.film_info(f"Creating data directory at {data_dir}")
            
            # Export project data
            project_data = {
                "archive_id": project.archive_id,
                "location": project.location,
                "location_code": project.location_code,
                "doc_type": project.doc_type,
                "project_path": str(project.project_path),
                "project_folder_name": project.project_folder_name,
                "document_folder_path": str(project.document_folder_path) if project.document_folder_path else None,
                "document_folder_name": project.document_folder_name,
                "has_oversized": project.has_oversized,
                "total_pages": project.total_pages,
                "total_pages_with_refs": project.total_pages_with_refs,
                "total_oversized": project.total_oversized,
                "documents_with_oversized": project.documents_with_oversized,
                "comlist_path": str(project.comlist_path) if project.comlist_path else None,
            }
            
            project_info_path = data_dir / f"{project.archive_id}_project_info.json"
            with open(project_info_path, "w", encoding="utf-8") as f:
                json.dump(project_data, f, indent=4)
                
            result_paths["project_info"] = project_info_path
            if self.logger:
                self.logger.film_info(f"Exported project info to {project_info_path}")
            
            # Export document data
            documents_data = []
            for doc in project.documents:
                doc_data = {
                    "doc_id": doc.doc_id,
                    "path": str(doc.path),
                    "pages": doc.pages,
                    "has_oversized": doc.has_oversized,
                    "total_oversized": doc.total_oversized,
                    "dimensions": doc.dimensions,
                    "ranges": doc.ranges,
                    "reference_pages": doc.reference_pages,
                    "total_references": doc.total_references,
                    "is_split": doc.is_split,
                    "roll_count": doc.roll_count,
                    "com_id": doc.com_id,
                    "total_pages_with_refs": doc.total_pages_with_refs
                }
                documents_data.append(doc_data)
            
            documents_path = data_dir / f"{project.archive_id}_documents.json"
            with open(documents_path, "w", encoding="utf-8") as f:
                json.dump(documents_data, f, indent=4)
                
            result_paths["documents"] = documents_path
            if self.logger:
                self.logger.film_info(f"Exported document data to {documents_path}")
            
            # Export film allocation if available
            if project.film_allocation:
                film_allocation_data = self._serialize_film_allocation(project.film_allocation)
                
                film_allocation_path = data_dir / f"{project.archive_id}_film_allocation.json"
                with open(film_allocation_path, "w", encoding="utf-8") as f:
                    json.dump(film_allocation_data, f, indent=4)
                    
                result_paths["film_allocation"] = film_allocation_path
                if self.logger:
                    self.logger.film_info(f"Exported film allocation to {film_allocation_path}")
                
                # Export film rolls data
                film_rolls_16mm = []
                for roll in project.film_allocation.rolls_16mm:
                    roll_data = {
                        "roll_id": roll.roll_id,
                        "film_type": roll.film_type.value,
                        "capacity": roll.capacity,
                        "pages_used": roll.pages_used,
                        "pages_remaining": roll.pages_remaining,
                        "has_split_documents": roll.has_split_documents,
                        "is_partial": roll.is_partial,
                        "remaining_capacity": roll.remaining_capacity,
                        "usable_capacity": roll.usable_capacity,
                        "creation_date": roll.creation_date,
                        "document_segments": [
                            {
                                "doc_id": segment.doc_id,
                                "path": str(segment.path),
                                "pages": segment.pages,
                                "page_range": segment.page_range,
                                "frame_range": segment.frame_range,
                                "document_index": segment.document_index,
                                "has_oversized": segment.has_oversized
                            }
                            for segment in roll.document_segments
                        ],
                        "document_ids": roll.document_ids,
                        "utilization": roll.utilization
                    }
                    film_rolls_16mm.append(roll_data)
                
                film_rolls_35mm = []
                for roll in project.film_allocation.rolls_35mm:
                    roll_data = {
                        "roll_id": roll.roll_id,
                        "film_type": roll.film_type.value,
                        "capacity": roll.capacity,
                        "pages_used": roll.pages_used,
                        "pages_remaining": roll.pages_remaining,
                        "has_split_documents": roll.has_split_documents,
                        "is_partial": roll.is_partial,
                        "remaining_capacity": roll.remaining_capacity,
                        "usable_capacity": roll.usable_capacity,
                        "creation_date": roll.creation_date,
                        "document_segments": [
                            {
                                "doc_id": segment.doc_id,
                                "path": str(segment.path),
                                "pages": segment.pages,
                                "page_range": segment.page_range,
                                "frame_range": segment.frame_range,
                                "document_index": segment.document_index,
                                "has_oversized": segment.has_oversized
                            }
                            for segment in roll.document_segments
                        ],
                        "document_ids": roll.document_ids,
                        "utilization": roll.utilization
                    }
                    film_rolls_35mm.append(roll_data)
                
                film_rolls_data = {
                    "rolls_16mm": film_rolls_16mm,
                    "rolls_35mm": film_rolls_35mm,
                    "statistics": {
                        "total_rolls_16mm": project.film_allocation.total_rolls_16mm,
                        "total_pages_16mm": project.film_allocation.total_pages_16mm,
                        "total_partial_rolls_16mm": project.film_allocation.total_partial_rolls_16mm,
                        "total_split_documents_16mm": project.film_allocation.total_split_documents_16mm,
                        "total_rolls_35mm": project.film_allocation.total_rolls_35mm,
                        "total_pages_35mm": project.film_allocation.total_pages_35mm,
                        "total_partial_rolls_35mm": project.film_allocation.total_partial_rolls_35mm,
                        "total_split_documents_35mm": project.film_allocation.total_split_documents_35mm
                    }
                }
                
                film_rolls_path = data_dir / f"{project.archive_id}_film_rolls.json"
                with open(film_rolls_path, "w", encoding="utf-8") as f:
                    json.dump(film_rolls_data, f, indent=4)
                    
                result_paths["film_rolls"] = film_rolls_path
                if self.logger:
                    self.logger.film_info(f"Exported film rolls to {film_rolls_path}")
            
            # Export index data if provided
            if index_data:
                index_path = data_dir / f"{project.archive_id}_index.json"
                with open(index_path, "w", encoding="utf-8") as f:
                    json.dump(index_data, f, indent=4)
                    
                result_paths["index"] = index_path
                if self.logger:
                    self.logger.film_info(f"Exported index data to {index_path}")
            
            if self.logger:
                self.logger.film_success(f"Successfully exported all results to {data_dir}")
                
        except Exception as e:
            if self.logger:
                self.logger.film_error(f"Error exporting results: {str(e)}")
            
        return result_paths
    
    def _serialize_film_allocation(self, film_allocation: FilmAllocation) -> Dict[str, Any]:
        """
        Serialize FilmAllocation object to a dictionary.
        
        Args:
            film_allocation: FilmAllocation object
            
        Returns:
            Serialized dictionary
        """
        return {
            "archive_id": film_allocation.archive_id,
            "project_name": film_allocation.project_name,
            "split_documents_16mm": film_allocation.split_documents_16mm,
            "split_documents_35mm": film_allocation.split_documents_35mm,
            "partial_rolls_16mm": film_allocation.partial_rolls_16mm,
            "partial_rolls_35mm": film_allocation.partial_rolls_35mm,
            "total_rolls_16mm": film_allocation.total_rolls_16mm,
            "total_pages_16mm": film_allocation.total_pages_16mm,
            "total_partial_rolls_16mm": film_allocation.total_partial_rolls_16mm,
            "total_split_documents_16mm": film_allocation.total_split_documents_16mm,
            "total_rolls_35mm": film_allocation.total_rolls_35mm,
            "total_pages_35mm": film_allocation.total_pages_35mm,
            "total_partial_rolls_35mm": film_allocation.total_partial_rolls_35mm,
            "total_split_documents_35mm": film_allocation.total_split_documents_35mm,
            "creation_date": film_allocation.creation_date,
            "version": film_allocation.version
        } 

File: übergabe/microexcel/microfilm/microfilmC/export_service.c
/**
 * @file export_service.c
 * @brief Implementation of export functions for the microfilm system
 */

#include "microfilm.h"

static MicrofilmError export_project_info(const Project *project, const char *output_dir);
static MicrofilmError export_documents_info(const Project *project, const char *output_dir);
static MicrofilmError export_film_allocation(const Project *project, const char *output_dir);
static MicrofilmError create_data_directory(const char *project_path, char *data_dir_path);
static void write_json_string(FILE *file, const char *key, const char *value, bool is_last);
static void write_json_int(FILE *file, const char *key, int value, bool is_last);
static void write_json_bool(FILE *file, const char *key, bool value, bool is_last);

MicrofilmError export_results(const Project *project, const char *output_dir, Logger *logger) {
    MicrofilmError error = {false, ""};
    
    if (!project) {
        error.has_error = true;
        strcpy(error.message, "Project is NULL");
        return error;
    }
    
    logger_section(logger, "Exporting Results");
    logger_log(logger, LOG_INFO, "EXPORT", "Starting export of project results");
    
    // Create data directory
    char data_dir[MAX_PATH_LENGTH];
    error = create_data_directory(project->project_path, data_dir);
    if (error.has_error) {
        return error;
    }
    
    logger_log(logger, LOG_INFO, "EXPORT", "Creating data directory at %s", data_dir);
    
    // Export project information
    error = export_project_info(project, data_dir);
    if (error.has_error) {
        logger_log(logger, LOG_ERROR, "EXPORT", "Failed to export project info: %s", error.message);
        return error;
    }
    
    logger_log(logger, LOG_INFO, "EXPORT", "Exported project info");
    
    // Export documents information
    error = export_documents_info(project, data_dir);
    if (error.has_error) {
        logger_log(logger, LOG_ERROR, "EXPORT", "Failed to export documents info: %s", error.message);
        return error;
    }
    
    logger_log(logger, LOG_INFO, "EXPORT", "Exported documents info");
    
    // Export film allocation if available
    if (project->film_allocation) {
        error = export_film_allocation(project, data_dir);
        if (error.has_error) {
            logger_log(logger, LOG_ERROR, "EXPORT", "Failed to export film allocation: %s", error.message);
            return error;
        }
        
        logger_log(logger, LOG_INFO, "EXPORT", "Exported film allocation");
    }
    
    logger_log(logger, LOG_SUCCESS, "EXPORT", "Successfully exported all project results to %s", data_dir);
    
    return error;
}

static MicrofilmError create_data_directory(const char *project_path, char *data_dir_path) {
    MicrofilmError error = {false, ""};
    
    snprintf(data_dir_path, MAX_PATH_LENGTH, "%s/.data", project_path);
    
    if (!directory_exists(data_dir_path)) {
        error = create_directory(data_dir_path);
        if (error.has_error) {
            return error;
        }
    }
    
    return error;
}

static MicrofilmError export_project_info(const Project *project, const char *output_dir) {
    MicrofilmError error = {false, ""};
    
    char file_path[MAX_PATH_LENGTH];
    snprintf(file_path, MAX_PATH_LENGTH, "%s/%s_project_info.json", output_dir, project->archive_id);
    
    FILE *file = fopen(file_path, "w");
    if (!file) {
        error.has_error = true;
        snprintf(error.message, MAX_ERROR_MESSAGE_LENGTH,
            "Failed to open file for writing: %s", file_path);
        return error;
    }
    
    fprintf(file, "{\n");
    write_json_string(file, "archive_id", project->archive_id, false);
    write_json_string(file, "location", project->location, false);
    write_json_string(file, "location_code", project_get_location_code(project), false);
    write_json_string(file, "doc_type", project->doc_type, false);
    write_json_string(file, "project_path", project->project_path, false);
    write_json_string(file, "project_folder_name", project->project_folder_name, false);
    
    if (strlen(project->document_folder_path) > 0) {
        write_json_string(file, "document_folder_path", project->document_folder_path, false);
    } else {
        fprintf(file, "  \"document_folder_path\": null,\n");
    }
    
    if (strlen(project->document_folder_name) > 0) {
        write_json_string(file, "document_folder_name", project->document_folder_name, false);
    } else {
        fprintf(file, "  \"document_folder_name\": null,\n");
    }
    
    write_json_bool(file, "has_oversized", project->has_oversized, false);
    write_json_int(file, "total_pages", project->total_pages, false);
    write_json_int(file, "total_pages_with_refs", project->total_pages_with_refs, false);
    write_json_int(file, "total_oversized", project->total_oversized, false);
    write_json_int(file, "documents_with_oversized", project->documents_with_oversized, false);
    
    if (strlen(project->comlist_path) > 0) {
        write_json_string(file, "comlist_path", project->comlist_path, true);
    } else {
        fprintf(file, "  \"comlist_path\": null\n");
    }
    
    fprintf(file, "}\n");
    
    fclose(file);
    return error;
}

static MicrofilmError export_documents_info(const Project *project, const char *output_dir) {
    MicrofilmError error = {false, ""};
    
    char file_path[MAX_PATH_LENGTH];
    snprintf(file_path, MAX_PATH_LENGTH, "%s/%s_documents.json", output_dir, project->archive_id);
    
    FILE *file = fopen(file_path, "w");
    if (!file) {
        error.has_error = true;
        snprintf(error.message, MAX_ERROR_MESSAGE_LENGTH,
            "Failed to open file for writing: %s", file_path);
        return error;
    }
    
    fprintf(file, "[\n");
    
    for (int i = 0; i < project->documents_count; i++) {
        const Document *doc = &project->documents[i];
        
        fprintf(file, "  {\n");
        write_json_string(file, "doc_id", doc->doc_id, false);
        write_json_string(file, "path", doc->path, false);
        write_json_int(file, "pages", doc->pages, false);
        write_json_bool(file, "has_oversized", doc->has_oversized, false);
        write_json_int(file, "total_oversized", doc->total_oversized, false);
        
        // Export dimensions array
        fprintf(file, "    \"dimensions\": [\n");
        for (int j = 0; j < doc->dimensions_count; j++) {
            const PageDimension *dim = &doc->dimensions[j];
            fprintf(file, "      [%.2f, %.2f, %d, %.2f]", 
                dim->width, dim->height, dim->page_index, dim->percent_over);
            if (j < doc->dimensions_count - 1) {
                fprintf(file, ",");
            }
            fprintf(file, "\n");
        }
        fprintf(file, "    ],\n");
        
        // Export ranges array
        fprintf(file, "    \"ranges\": [\n");
        for (int j = 0; j < doc->ranges_count; j++) {
            const PageRange *range = &doc->ranges[j];
            fprintf(file, "      [%d, %d]", range->start, range->end);
            if (j < doc->ranges_count - 1) {
                fprintf(file, ",");
            }
            fprintf(file, "\n");
        }
        fprintf(file, "    ],\n");
        
        // Export reference pages array
        fprintf(file, "    \"reference_pages\": [\n");
        for (int j = 0; j < doc->reference_pages_count; j++) {
            fprintf(file, "      %d", doc->reference_pages[j]);
            if (j < doc->reference_pages_count - 1) {
                fprintf(file, ",");
            }
            fprintf(file, "\n");
        }
        fprintf(file, "    ],\n");
        
        write_json_int(file, "total_references", doc->total_references, false);
        write_json_bool(file, "is_split", doc->is_split, false);
        write_json_int(file, "roll_count", doc->roll_count, false);
        write_json_int(file, "com_id", doc->com_id, false);
        write_json_int(file, "total_pages_with_refs", doc->pages + doc->total_references, true);
        
        fprintf(file, "  }");
        if (i < project->documents_count - 1) {
            fprintf(file, ",");
        }
        fprintf(file, "\n");
    }
    
    fprintf(file, "]\n");
    
    fclose(file);
    return error;
}

static MicrofilmError export_film_allocation(const Project *project, const char *output_dir) {
    MicrofilmError error = {false, ""};
    
    const FilmAllocation *allocation = project->film_allocation;
    if (!allocation) {
        return error;
    }
    
    char file_path[MAX_PATH_LENGTH];
    snprintf(file_path, MAX_PATH_LENGTH, "%s/%s_film_allocation.json", output_dir, project->archive_id);
    
    FILE *file = fopen(file_path, "w");
    if (!file) {
        error.has_error = true;
        snprintf(error.message, MAX_ERROR_MESSAGE_LENGTH,
            "Failed to open file for writing: %s", file_path);
        return error;
    }
    
    fprintf(file, "{\n");
    write_json_string(file, "archive_id", allocation->archive_id, false);
    write_json_string(file, "project_name", allocation->project_name, false);
    
    // Export 16mm rolls
    fprintf(file, "  \"rolls_16mm\": [\n");
    for (int i = 0; i < allocation->rolls_16mm_count; i++) {
        const FilmRoll *roll = &allocation->rolls_16mm[i];
        
        fprintf(file, "    {\n");
        write_json_int(file, "roll_id", roll->roll_id, false);
        write_json_string(file, "film_type", "16mm", false);
        write_json_int(file, "capacity", roll->capacity, false);
        write_json_int(file, "pages_used", roll->pages_used, false);
        write_json_int(file, "pages_remaining", roll->pages_remaining, false);
        write_json_string(file, "film_number", roll->film_number, false);
        write_json_string(file, "status", roll->status, false);
        write_json_bool(file, "has_split_documents", roll->has_split_documents, false);
        write_json_bool(file, "is_partial", roll->is_partial, false);
        write_json_int(file, "remaining_capacity", roll->remaining_capacity, false);
        write_json_int(file, "usable_capacity", roll->usable_capacity, false);
        write_json_string(file, "creation_date", roll->creation_date, false);
        
        // Export document segments
        fprintf(file, "      \"document_segments\": [\n");
        for (int j = 0; j < roll->segments_count; j++) {
            const DocumentSegment *segment = &roll->document_segments[j];
            
            fprintf(file, "        {\n");
            write_json_string(file, "doc_id", segment->doc_id, false);
            write_json_string(file, "path", segment->path, false);
            write_json_int(file, "pages", segment->pages, false);
            fprintf(file, "          \"page_range\": [%d, %d],\n", 
                segment->page_range.start, segment->page_range.end);
            fprintf(file, "          \"frame_range\": [%d, %d],\n", 
                segment->frame_range.start, segment->frame_range.end);
            write_json_int(file, "document_index", segment->document_index, false);
            write_json_bool(file, "has_oversized", segment->has_oversized, true);
            fprintf(file, "        }");
            if (j < roll->segments_count - 1) {
                fprintf(file, ",");
            }
            fprintf(file, "\n");
        }
        fprintf(file, "      ]\n");
        
        fprintf(file, "    }");
        if (i < allocation->rolls_16mm_count - 1) {
            fprintf(file, ",");
        }
        fprintf(file, "\n");
    }
    fprintf(file, "  ],\n");
    
    // Export 35mm rolls
    fprintf(file, "  \"rolls_35mm\": [\n");
    for (int i = 0; i < allocation->rolls_35mm_count; i++) {
        const FilmRoll *roll = &allocation->rolls_35mm[i];
        
        fprintf(file, "    {\n");
        write_json_int(file, "roll_id", roll->roll_id, false);
        write_json_string(file, "film_type", "35mm", false);
        write_json_int(file, "capacity", roll->capacity, false);
        write_json_int(file, "pages_used", roll->pages_used, false);
        write_json_int(file, "pages_remaining", roll->pages_remaining, false);
        write_json_string(file, "film_number", roll->film_number, false);
        write_json_string(file, "status", roll->status, false);
        write_json_bool(file, "has_split_documents", roll->has_split_documents, false);
        write_json_bool(file, "is_partial", roll->is_partial, false);
        write_json_int(file, "remaining_capacity", roll->remaining_capacity, false);
        write_json_int(file, "usable_capacity", roll->usable_capacity, false);
        write_json_string(file, "creation_date", roll->creation_date, false);
        
        // Export document segments
        fprintf(file, "      \"document_segments\": [\n");
        for (int j = 0; j < roll->segments_count; j++) {
            const DocumentSegment *segment = &roll->document_segments[j];
            
            fprintf(file, "        {\n");
            write_json_string(file, "doc_id", segment->doc_id, false);
            write_json_string(file, "path", segment->path, false);
            write_json_int(file, "pages", segment->pages, false);
            fprintf(file, "          \"page_range\": [%d, %d],\n", 
                segment->page_range.start, segment->page_range.end);
            fprintf(file, "          \"frame_range\": [%d, %d],\n", 
                segment->frame_range.start, segment->frame_range.end);
            write_json_int(file, "document_index", segment->document_index, false);
            write_json_bool(file, "has_oversized", segment->has_oversized, true);
            fprintf(file, "        }");
            if (j < roll->segments_count - 1) {
                fprintf(file, ",");
            }
            fprintf(file, "\n");
        }
        fprintf(file, "      ]\n");
        
        fprintf(file, "    }");
        if (i < allocation->rolls_35mm_count - 1) {
            fprintf(file, ",");
        }
        fprintf(file, "\n");
    }
    fprintf(file, "  ],\n");
    
    // Export statistics
    write_json_int(file, "total_rolls_16mm", allocation->total_rolls_16mm, false);
    write_json_int(file, "total_pages_16mm", allocation->total_pages_16mm, false);
    write_json_int(file, "total_partial_rolls_16mm", allocation->total_partial_rolls_16mm, false);
    write_json_int(file, "total_split_documents_16mm", allocation->total_split_documents_16mm, false);
    write_json_int(file, "total_rolls_35mm", allocation->total_rolls_35mm, false);
    write_json_int(file, "total_pages_35mm", allocation->total_pages_35mm, false);
    write_json_int(file, "total_partial_rolls_35mm", allocation->total_partial_rolls_35mm, false);
    write_json_int(file, "total_split_documents_35mm", allocation->total_split_documents_35mm, false);
    write_json_string(file, "creation_date", allocation->creation_date, false);
    write_json_string(file, "version", allocation->version, true);
    
    fprintf(file, "}\n");
    
    fclose(file);
    return error;
}

static void write_json_string(FILE *file, const char *key, const char *value, bool is_last) {
    fprintf(file, "  \"%s\": \"%s\"", key, value);
    if (!is_last) {
        fprintf(file, ",");
    }
    fprintf(file, "\n");
}

static void write_json_int(FILE *file, const char *key, int value, bool is_last) {
    fprintf(file, "  \"%s\": %d", key, value);
    if (!is_last) {
        fprintf(file, ",");
    }
    fprintf(file, "\n");
}

static void write_json_bool(FILE *file, const char *key, bool value, bool is_last) {
    fprintf(file, "  \"%s\": %s", key, value ? "true" : "false");
    if (!is_last) {
        fprintf(file, ",");
    }
    fprintf(file, "\n");
}

